### PR TITLE
feat(hal-core): address type improvements

### PR DESCRIPTION
**Pull request by hawkw**
This branch makes the following changes:

* feat(hal-core): implement Address::{align_up, align_down}

  Add generic default implementations of `align_up` and `align_down` to
  all types implementing Address. These are implemented using a new 
  required `from_usize` constructor.

* feat(hal-core): `Address::as_ptr` returns typed mut ptr

  As discussed in [this comment][1], `as_ptr` now returns a `*mut T`
  rather than a `*const ()`. It will now assert that the address is
  correctly aligned for a `T`, using a new `is_aligned_for::<T>` 
  convenience method.

* feat(hal-core): validate/canonicalize x86_64 addrs 

  Add `#[cfg(target_arch = "x86_64")]` implementations of 
  `Address::from_usize` for `PAddr`/`VAddr` that assert the address
  is sign-extended/canonicalized when compiled in debug mode. 
  Additionally, add `from_usize_checked` constructors that return
  `Result` and always validate that the constructed addresses are 
   valid.

* feat(hal-core): add inherent versions of address trait methods

### CODE REVIEW

File: hal-core/src/addr.rs [hunk 1]
```diff
@@ -13,16 +13,40 @@ pub trait Address:
...
+    ///
+    /// # Panics
+    ///
+    /// - If `align` is not a power of two.
+    fn align_up<A: Into<usize>>(self, align: A) -> Self {
+        let align = align.into();
+        assert!(align.is_power_of_two());
+        let aligned = self.as_usize() & !(align - 1);
+        Self::from_usize(aligned)
+    }
```
[the remark this answers is not in the document]
Comment (hawkw) [reply hal-core/src/addr.rs]: omg, this is my fault — this is actually the implementation of `align_down`...I had implemented these methods on another branch and copy-pasted them over, and it looks like i swapped them by mistake. my bad.

File: hal-core/src/addr.rs [hunk 2]
```diff
@@ -132,32 +169,28 @@ impl Address for PAddr {
...
-    }
-
-    /// Returns `true` if `self` is aligned on the specified alignment.
-    fn is_aligned<A: Into<usize>>(self, _align: A) -> bool {
-        unimplemented!("eliza")
-    }
-
-    fn as_ptr(&self) -> *const () {
-        unimplemented!("eliza")
+    #[cfg(not(target_arch = "x86_64"))]
```
[the remark this answers is not in the document]
Comment (hawkw) [reply hal-core/src/addr.rs]: platforms that don't have additional logic for what is considered a valid address can just use the un-cfg'd implementation.

[the remark this answers is not in the document]
Comment (mystor) [reply hal-core/src/addr.rs]: Shouldn't that implementation at least use `u` as the value, rather than just `0`?

`Self(0)` as the value for all addresses seems like very surprising behaviour.

File: hal-core/src/addr.rs [hunk 3]
```diff
@@ -218,32 +251,25 @@ impl Address for VAddr {
...
-    }
-
-    /// Returns `true` if `self` is aligned on the specified alignment.
-    fn is_aligned<A: Into<usize>>(self, _align: A) -> bool {
-        unimplemented!("eliza")
+    #[inline]
+    #[cfg(target_arch = "x86_64")]
+    fn from_usize(u: usize) -> Self {
+        // sign extend bit 47
+        let value = ((u as i64) << 16) as u64 >> 16;
```
[the remark this answers is not in the document]
Comment (hawkw) [reply hal-core/src/addr.rs]: whoops, fixed

File: hal-core/src/addr.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (hawkw) [reply hal-core/src/addr.rs]: ag, it was supposed to. OOPS!

File: hal-core/src/addr.rs [hunk 4]
```diff
@@ -13,16 +13,40 @@ pub trait Address:
...
+    /// - If `align` is not a power of two.
+    fn align_up<A: Into<usize>>(self, align: A) -> Self {
+        let align = align.into();
+        assert!(align.is_power_of_two());
+        let mask = align - 1;
+        let u = self.as_usize();
+        if u & mask == 0 {
+            return self;
+        }
+        let aligned = (u | mask) + 1;
```
[the remark this answers is not in the document]
Comment (hawkw) [reply hal-core/src/addr.rs]: idk, we'd have to test it.

[the remark this answers is not in the document]
Comment (hawkw) [reply hal-core/src/addr.rs]: (but probably very slightly)

File: hal-core/src/addr.rs [hunk 5]
```diff
@@ -132,32 +169,28 @@ impl Address for PAddr {
...
-        unimplemented!("eliza")
+    #[inline]
+    #[cfg(target_arch = "x86_64")]
+    fn from_usize(u: usize) -> Self {
+        const MASK: usize = 0xFFF0_0000_0000_0000;
+        debug_assert!(
+            u & MASK == 0,
+            "x86_64 physical addresses may not have the 12 most significant bits set!"
+        );
+        Self(u & !MASK)
```
[the remark this answers is not in the document]
Comment (hawkw) [reply hal-core/src/addr.rs]: hmm, no, it would just fault. that's a good point.

File: hal-core/src/addr.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mystor) [reply hal-core/src/addr.rs]: that's fair. Let's stick with this.

**Comment by hawkw** [comment]
changed the `from_usize` impls to debug_assert validity, and let the CPU fault in release mode. we probably want separate constructors that _do_ make the address valid?

File: hal-core/src/addr.rs [hunk 6]
```diff
@@ -218,32 +250,28 @@ impl Address for VAddr {
...
-        unimplemented!("eliza")
-    }
-
-    /// Returns `true` if `self` is aligned on the specified alignment.
-    fn is_aligned<A: Into<usize>>(self, _align: A) -> bool {
-        unimplemented!("eliza")
+    #[inline]
+    #[cfg(target_arch = "x86_64")]
+    fn from_usize(u: usize) -> Self {
+        debug_assert!(
```
Comment (mystor) [inline hal-core/src/addr.rs]: Should this be a `debug_assert_eq!`?

File: hal-core/src/addr.rs [hunk 7]
```diff
@@ -13,16 +13,40 @@ pub trait Address:
...
 
     /// Aligns `self` down to `align`.
     ///
     /// The specified alignment must be a power of two.
-    fn align_down<A: Into<usize>>(self, align: A) -> Self;
+    ///
+    /// # Panics
+    ///
+    /// - If `align` is not a power of two.
+    fn align_down<A: Into<usize>>(self, align: A) -> Self {
```
Comment (mystor) [inline hal-core/src/addr.rs]: We can totally do this in a different patch, but it might be nice to have `align_up_for` and `align_down_for` helper methods like the `is_aligned_for` method.

File: hal-core/src/addr.rs [hunk 8]
```diff
@@ -132,32 +169,27 @@ impl Address for PAddr {
...
 
-    fn align_up<A: Into<usize>>(self, _align: A) -> Self {
-        unimplemented!("eliza")
+    #[inline]
+    #[cfg(target_arch = "x86_64")]
+    fn from_usize(u: usize) -> Self {
+        debug_assert!(
+            u & MASK == 0,
+            "x86_64 physical addresses may not have the 12 most significant bits set!"
+        );
```
Comment (mystor) [inline hal-core/src/addr.rs]: Might be worth `#[cfg(target_arch = "x86_64")]`-ing only this assertion, and keeping the rest of the `from_usize` impl common?

File: hal-core/src/addr.rs [hunk 9]
```diff
@@ -50,7 +74,20 @@ pub trait Address:
...
+    #[inline]
+    fn is_aligned_for<T>(self) -> bool {
+        self.is_aligned(core::mem::align_of::<T>())
+    }
+
+    /// # Panics
+    ///
+    /// - If `self` is not aligned for a `T`-typed value.
+    fn as_ptr<T>(self) -> *mut T {
+        assert!(self.is_aligned_for::<T>());
```
Comment (mystor) [inline hal-core/src/addr.rs]: I suppose we need this restriction for rust, even though there's no aligned read requirement on x64?

File: hal-core/src/addr.rs [hunk 10]
```diff
@@ -260,3 +288,58 @@ impl fmt::Debug for VAddr {
...
+    //         actual, expected
+    //     );
+    //     let actual = VAddr::from_usize(123 | (1010 << 47)).as_usize();
+    //     let expected = 123;
+    //     assert_eq!(
+    //         actual, expected,
+    //         "\n  left: `{:064b}`\n right: `{:064b}`",
+    //         actual, expected
+    //     );
+    // }
```
Comment (mystor) [inline hal-core/src/addr.rs]: We probably want to remove this.

File: hal-core/src/addr.rs (hunk 9 above)
Comment (hawkw) [reply hal-core/src/addr.rs]: that's correct, will add a note
